2D Simulation/movesTracker.js

```diff
@@ -1,99 +1,114 @@
-            + export_block_action(action) + " "
-            + action.end_date + " "
-            + export_position(action.left_pos, action.top_pos) + " "
-            + export_position(action.new_left_pos, action.new_top_pos);
-    }
-
-    function export_position(left_pos, top_pos) {
-        return "(" + left_pos + "," + top_pos + ")";
-    }
-
-    function export_block_action(action) {
-        return "Block id: " + action.id + " "
-            + "Letter: " + action.letter + " "
-            + "Color: " + action.color;
+        this.actions.push(new _Move(id, letter, color,
+                                    orig_left, orig_top,
+                                    end_left, end_top));
     }
 
     this.export_actions = function () {
         let action_strs = [];
 
         for (let action of this.actions) {
-            action_strs.push(action.export_func(action));
+            action_strs.push(action.export_to_string());
         }
 
         return action_strs;
     }
 }
+
+function _Gesture(left_pos_percent, top_pos_percent) {
+    this.end_date = getDateTime();
+    this.left_pos = left_pos_percent;
+    this.top_pos = top_pos_percent;
+
+    this.export_to_string = function () {
+        return "Gesture" + " "
+            + this.end_date + " "
+            + _export_position(this.left_pos, this.top_pos);
+    };
+}
+
+function _Instruction(text) {
+    this.end_date = getDateTime();
+    this.text = text;
+
+    this.export_to_string = function () {
+        return "Instruction" + " "
+            + this.end_date + " "
+            + this.text;
+    };
+}
+
+function _Flip(id, letter, color, x_pos, y_pos) {
+    this.id = id;
+    this.letter = letter;
+    this.color = color;
+    this.end_date = getDateTime();
+    this.left_pos = x_pos;
+    this.top_pos = y_pos;
+
+    this.export_to_string = function () {
+        return "Flip" + " "
+            + _export_block_action(this) + " "
+            + this.end_date + " "
+            + _export_position(this.left_pos, this.top_pos);
+    }
+}
+
+function _Move(id, letter, color, orig_left, orig_top, end_left, end_top) {
+    this.id = id;
+    this.letter = letter;
+    this.color = color;
+    this.end_date = getDateTime();
+    this.left_pos = orig_left;
+    this.top_pos = orig_top;
+    this.new_left_pos = end_left;
+    this.new_top_pos = end_top;
+
+    this.export_to_string = function () {
+        return "Movement" + " "
+            + _export_block_action(this) + " "
+            + this.end_date + " "
+            + _export_position(this.left_pos, this.top_pos) + " "
+            + _export_position(this.new_left_pos, this.new_top_pos);
+    };
+}
+
+function _export_block_action(action) {
+    return "Block id: " + action.id + " "
+        + "Letter: " + action.letter + " "
+        + "Color: " + action.color;
+}
+
+function _export_position(left_pos, top_pos) {
+    return "(" + left_pos + "," + top_pos + ")";
+}
+
+if (typeof module !== 'undefined'
```

Comment:
As long as we're doing stuff like this would it make sense to use actual js modules? We would probably need to add in babel or something

Reply:
SIgh... You have no idea how much I'd like to do that.

First of all, setting everything up to use modules would take a good deal of refactoring effort. I think it would be worth it, but I want to push through the E2E tests before we do anything that exciting.

Secondly, for whatever reason, Jasmine does not like `import` / `export` keywords (it says unrecognized token or something like that), so yes, we'd have to use babel for that, which I haven't tested much with yet and as such still might not cooperate.

The last option might be RequireJS, which Jasmine also does not seem to appreciate very much when run through the command line. Instead, you have to use their `SpecRunner.html` and load RequireJS into it in order for it to work...
 
Also of note is that the HTML would no longer have access to global functions (like the various buttons wouldn't see their callbacks anymore), and as such the JavaScript would have to use jQuery to attach the functions to the buttons manually. I still think this is worthwhile, mind, it's just that this will take a concerted effort.

TL;DR: Hopefully soon?